Better support for Symbol, null, undefined, Infinity, NaN

maxwellgerber opened the pull request:
Improves the `defaultCacheKey` function to handle types that don't play well with `JSON.stringify` using the approach discussed in https://github.com/sindresorhus/mem/issues/20

Problems I ran into - 
It's impossible to set a symbol as a key in a `WeakMap`, so I had to use a Map instead

```
m = new WeakMap();
m.set(Symbol())
// Uncaught TypeError: Invalid value used as weak map key
```
I don't foresee garbage collection being an issue for two reasons
1. In my experience people generally don't create symbols dynamically, instead they're used as enums at program start
1. Based on [this SO Post](https://stackoverflow.com/questions/39997034/are-es6-global-symbols-garbage-collected) it looks like symbols may not be garbage collected anyway

closes #20

maxwellgerber commented:
Appears `SharedArrayBuffer` is not supported in Node 6. Do you have a preference for how to approach handling version discrepancies like that? I can wrap in a try/catch otherwise.

maxwellgerber commented:
@sindresorhus done.
Now that I look at it, I really don't like the behavior for mutable types.
```
let map = new Map();
map.set('a', 1);
let memoized = mem(obj => obj.get('a'));
memoized(map) // 1
map.set('a', 2);
memoized(map) // 1
```
Unfortunately, pretty much everything is mutable, even regexes.
```
let reg = /asdf/
reg.toString() // '/asdf/'
reg.compile()
reg.toString() // '/(?:)/'
```
Maybe it would be better to only use the `ValueHandlers` and `symbolMap` from this PR and drop the prototype stuff.

GrosSacASac commented:
@maxwellgerber there is a problem with multiple values, change the test 'memoize with null and undefined' to this and see test failing

```
test('memoize with null and undefined', t => {
	let i = 0;
	const memoized = m(() => i++);
	t.is(memoized(), 0);
	t.is(memoized(), 0);
	t.is(memoized(null), 1);
	t.is(memoized(null), 1);
	t.is(memoized(undefined), 2);
	t.is(memoized(undefined), 2);
	t.is(memoized([null, undefined]), 3);
	t.is(memoized([null, undefined]), 3);
	t.is(memoized([undefined, null]), 4);
	t.is(memoized([undefined, null]), 4);
	t.is(memoized("undefined"), 5);
	t.is(memoized("undefined"), 5);
	t.is(memoized("null"), 6);
	t.is(memoized("null"), 6);
	t.is(memoized(null, null), 7);
	t.is(memoized(null, null), 7);
	t.is(memoized("null", "null"), 8);
	t.is(memoized("null", "null"), 8);
});
```

GrosSacASac commented on test.js:
```diff
@@ -28,6 +33,145 @@ test('memoize with multiple non-primitive arguments', t => {
 	t.is(memoized({foo: true}, {bar: false}, {baz: true}), 2);
 });
 
+test('memoize with null and undefined', t => {
```
see comment in discussion

GrosSacASac replied:
ok

maxwellgerber commented:
@GrosSacASac Great catch! 
Fixed, and tests updated. 

I've dropped the stuff about handling all builtin objects - this PR will only address Symbols and values that don't serialize well (`null`, `undefined`, `Infinity`, `NaN`) 

Some documentation and tests RE the Regex issue have been added.

@sindresorhus I think this guy is ready for review

sindresorhus commented on index.js:
```diff
@@ -1,9 +1,28 @@
+	}
+
+	if (typeof value === 'symbol') {
+		TypesMap.set(value, uuid());
```
@szmarczak Can you elaborate on why it's a memory leak?

szmarczak replied:
Because it is never deleted.

maxwellgerber replied:
see: https://github.com/tc39/ecma262/issues/1194
I guess you can cause a problem if someone creates hundreds of thousands of symbols, but I can't imagine why anyone would ever do that.

@bfred-it's ManyKeysMap in #33 does the same thing.
https://github.com/bfred-it/many-keys-map/blob/v1.0.1/index.js#L22

What should I be doing instead?

szmarczak replied:
You're right. I think you should work together with @bfred-it - this topic isn't as easy as it seems.

bfred-it replied:
There's no way around it, cached objects (or in this case just Symbols) will stay around until the cache is `.clear`ed

bfred-it commented:
I'd like your thoughts on #33. It would fix this issue once for all, but it requires two changes to how `mem` works.

bfred-it commented:
I think all we need here is a reliable and comprehensive JSON replacer module, rather than interlacing this code with `mem`

maxwellgerber commented:
Closing for inactivity - hope #43 make it in
